File: go/internal/tools/system.go
package tools

import (
	"fmt"
	"runtime"
	"time"
)

var startTime = time.Now()

// SystemInfo returns system information
func SystemInfo(args map[string]interface{}) (interface{}, error) {
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)

	return map[string]interface{}{
		"platform":      runtime.GOOS,
		"architecture":  runtime.GOARCH,
		"goVersion":     runtime.Version(),
		"numCPU":        runtime.NumCPU(),
		"numGoroutine":  runtime.NumGoroutine(),
		"uptime":        time.Since(startTime).String(),
		"memory": map[string]interface{}{
			"alloc":      fmt.Sprintf("%d MB", memStats.Alloc/1024/1024),
			"totalAlloc": fmt.Sprintf("%d MB", memStats.TotalAlloc/1024/1024),
			"sys":        fmt.Sprintf("%d MB", memStats.Sys/1024/1024),
			"numGC":      memStats.NumGC,
		},
		"serverType": "Go MCP Server",
		"timestamp":  time.Now().Format(time.RFC3339),
	}, nil
}

// GetSystemInfoSchema returns the JSON schema for system_info
func GetSystemInfoSchema() map[string]interface{} {
	return map[string]interface{}{
		"type":       "object",
		"properties": map[string]interface{}{},
	}
}

File: rust/mcp-core/src/tools/system.rs
use super::ToolHandler;
use crate::types::Tool;
use anyhow::Result;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::time::SystemTime;

lazy_static::lazy_static! {
    static ref START_TIME: SystemTime = SystemTime::now();
}

pub struct SystemInfoTool;

#[async_trait::async_trait]
impl ToolHandler for SystemInfoTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "system_info".to_string(),
            description: "Get system information about the Rust runtime".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {}
            }),
        }
    }

    async fn execute(&self, _args: HashMap<String, Value>) -> Result<Value> {
        let uptime = START_TIME.elapsed().unwrap_or_default();

        Ok(json!({
            "platform": std::env::consts::OS,
            "architecture": std::env::consts::ARCH,
            "rustVersion": env!("CARGO_PKG_RUST_VERSION", "unknown"),
            "uptime": format!("{:?}", uptime),
            "serverType": "Rust MCP Server",
            "timestamp": chrono::Utc::now().to_rfc3339(),
        }))
    }
}
